applications/serial_lte_modem/src/slm_ppp.c

```diff
@@ -0,0 +1,311 @@
+
+		if (inet_pton(AF_INET6, addr6, &in6) != 1)
+			return false;
+
+		/* The interface identifier is the last 64 bits of the IPv6 address. */
+		BUILD_ASSERT(sizeof(in6) >= sizeof(opts->iid));
+		/* Setting it doesn't seem to change anything in practice, but this
+		 * is what it's meant for, so fill it anyway.
+		 */
+		memcpy(opts->iid, (uint8_t *)(&in6 + 1) - sizeof(opts->iid), sizeof(opts->iid));
+	}
+	return true;
+}
+
+static int slm_ppp_start(void)
+{
+	int ret;
+	unsigned int mtu;
+	struct ppp_context *const ctx = net_if_l2_data(ppp_iface);
+
+	if (!configure_ppp_link_ip_addresses(ctx)) {
+		return -EADDRNOTAVAIL;
+	}
+
+	ret = pdn_dynamic_params_get(PDP_CID, &ctx->ipcp.my_options.dns1_address,
+				     &ctx->ipcp.my_options.dns2_address, &mtu);
+	if (ret) {
+		return ret;
+	}
+
+	/* Set the PPP MTU to that of the LTE link. */
+	mtu = MIN(mtu, sizeof(ppp_data_buf));
+	net_if_set_mtu(ppp_iface, mtu);
+	LOG_DBG("MTU set to %u.", mtu);
+
+	ret = net_if_up(ppp_iface);
+	if (ret) {
+		LOG_ERR("Failed to bring PPP interface up (%d).", ret);
+		return ret;
+	}
+
+	if (!open_ppp_sockets()) {
+		close_ppp_sockets();
+		net_if_down(ppp_iface);
+		return -ENOTCONN;
+	}
+
+	LOG_INF("PPP started.");
+
+	k_thread_create(&ppp_data_passing_thread_id, ppp_data_passing_thread_stack,
+			K_THREAD_STACK_SIZEOF(ppp_data_passing_thread_stack),
+			ppp_data_passing_thread, NULL, NULL, NULL,
+			K_PRIO_COOP(10), 0, K_NO_WAIT);
+	k_thread_name_set(&ppp_data_passing_thread_id, "ppp_data_passing");
+
+	return 0;
+}
+
+/** @return Whether PPP is fully started. */
+static bool slm_ppp_is_started(void)
+{
+	return net_if_flag_is_set(ppp_iface, NET_IF_UP)
+	    && ppp_fds[ZEPHYR_FD_IDX] >= 0
+	    && ppp_fds[MODEM_FD_IDX] >= 0;
+}
+
+static void slm_ppp_stop(void)
+{
+	if (!slm_ppp_is_started()) {
+		return;
+	}
+	/* First bring the interface down so that slm_ppp_is_started()
+	 * returns false right away. This is to prevent trying to stop PPP at the
+	 * same time from multiple sources: the original one (e.g. "AT#XPPP=0")
+	 * and the data passing thread. The latter attempts to stop PPP when it receives
+	 * an error on some of the sockets, which happens when they are closed.
+	 */
+	const int ret = net_if_down(ppp_iface);
+
+	if (ret) {
+		LOG_WRN("Failed to bring PPP interface down (%d).", ret);
+	}
+
+	close_ppp_sockets();
+
+	k_thread_join(&ppp_data_passing_thread_id, K_SECONDS(1));
+
+	LOG_INF("PPP stopped.");
+}
+
+int slm_ppp_init(void)
+{
+	if (!device_is_ready(ppp_uart_dev)) {
+		return -EAGAIN;
+	}
+
+	ppp_iface_idx = net_if_get_by_name(CONFIG_NET_PPP_DRV_NAME);
+	ppp_iface = net_if_get_by_index(ppp_iface_idx);
+	if (!ppp_iface) {
+		LOG_ERR("PPP network interface not found (%d).", ppp_iface_idx);
+		return -ENODEV;
+	}
+	net_if_flag_set(ppp_iface, NET_IF_POINTOPOINT);
+
+	LOG_DBG("PPP initialized.");
+	return 0;
+}
+
+/* Handles AT#XPPP commands. */
+int handle_at_ppp(enum at_cmd_type cmd_type)
+{
+	if (cmd_type != AT_CMD_TYPE_SET_COMMAND
```

Comment:
I'd move the check after the variable declarations, for me it looks a bit awkward that way.